### PR TITLE
[wip] implement FILL_COST

### PR DESCRIPTION
**Pull request by gballet**
Extract the information of whether the value was present from `StateDB` and use it to let the access witness know if the fill costs are to be charged.

### CODE REVIEW

File: core/state/state_object.go [hunk 1]
```diff
@@ -203,6 +206,8 @@ func (s *stateObject) GetCommittedState(key common.Hash) common.Hash {
 				s.db.setError(err)
 			}
 			value.SetBytes(content)
+		} else {
+			s.isFilled[key] = true // if we get here, it means that the fill wasn't cached and so it's a fill
```
Comment (jsign) [inline core/state/state_object.go:210]: Any storage-slot write _must_ have called `GetCommitedState(...)` before for this map to be updated properly for a subsequent _further_ witness recording. If in any current (or future) case this doesn't hold, the gas will be accounted incorrectly.

Isn't this quite dangerous assumption?

File: core/state/state_object.go [hunk 2]
```diff
@@ -242,6 +252,7 @@ func (s *stateObject) SetState(key, value common.Hash) {
 		prevalue: prev,
 	})
 	s.setState(key, value)
+	s.isFilled[key] = false
```
Comment (jsign) [inline core/state/state_object.go:255]: Some questions:
- Shouldn't this also happen before L246?
- This strategy seems to assume any witness record is done before a `SetState` call. If this operations are swapped, then we'd be doing the wrong thing? If that's the case sounds like an implicit/dangerous assumption.

File: core/vm/operations_verkle.go [hunk 3]
```diff
@@ -23,15 +23,16 @@ import (
 )
```
Comment (jsign) [inline core/vm/operations_verkle.go:23]: Maybe this is due to the PR being WIP, but we should also mark as `isFill` any case where a new account is created.

File: core/state/access_witness.go [hunk 4]
```diff
@@ -92,40 +92,40 @@ func (aw *AccessWitness) Copy() *AccessWitness {
 func (aw *AccessWitness) TouchFullAccount(addr []byte, isWrite bool) uint64 {
```
Comment (jsign) [inline core/state/access_witness.go:92]: I wonder if might be better to let the `AccessWitness` have access to `StateDB` to resolve if a write is or isn't filling.

As in, any time an `isWrite` (which is an existing parameter in stuff) has to charge gas, then ask `StateDB`: "Is this empty in the state?", if "Yes" charge, if "No" don't.

This avoids this new `isFill` parameter, avoids all callers having to send the right value, and maybe other risks?

Sounds like a natural design but maybe you've discarded that for reasons.
But look the further comments I did in below to see some risks I see with the current strategy.

File: core/state/state_object.go [hunk 5]
```diff
@@ -112,6 +114,7 @@ func newObject(db *StateDB, address common.Address, acct *types.StateAccount) *s
 		originStorage:  make(Storage),
 		pendingStorage: make(Storage),
 		dirtyStorage:   make(Storage),
+		isFilled:       make(map[common.Hash]bool),
```
Comment (jsign) [inline core/state/state_object.go:117]: This is a nit irrelevant for now but could be `map[common.Hash]struct{}` to save space. There's no need to distinguish "empty entry" form "false" value.

File: core/state/state_object.go [hunk 6]
```diff
@@ -242,6 +252,7 @@ func (s *stateObject) SetState(key, value common.Hash) {
 		prevalue: prev,
 	})
 	s.setState(key, value)
+	s.isFilled[key] = false
 }
```
Comment (jsign) [inline core/state/state_object.go:257]: General comment, and might be part of WIP: sounds like this `isFilled` map should be journal-compatible to handle reverts/reorgs/persistence?

File: core/vm/operations_verkle.go (hunk 3 above)
Comment (gballet) [reply core/vm/operations_verkle.go:23]: yeah this version is only about SSTORE for now, in order to validate a few things while keeping stuff simple.

**Comment by gballet** [comment]
We're going a different route with Gary's changes, closing this one.

**Pull request closed without merging**